src/operator/subgraph/subgraph_property.h

```diff
@@ -200,7 +197,7 @@ typedef dmlc::ThreadLocalStore<std::unordered_map<std::string, std::unordered_se
 
 #define MXNET_REGISTER_SUBGRAPH_PROPERTY(Name, SubgraphPropertyType) \
   static DMLC_ATTRIBUTE_UNUSED auto __make_ ## SubgraphPropertyType ## _ ## Name ## __ = \
-    SubgraphPropertyRegistry::Get()->__REGISTER_OR_GET__(#Name, &SubgraphPropertyType::Create)
+    SubgraphPropertyRegistry::Get()->__REGISTER__(#Name, &SubgraphPropertyType::Create)
```

Comment:
Any reason of changing this? I remember there was a discussion in the first subgraph API PR and using `__REGISTER__` may result in terminating the program.
https://github.com/apache/incubator-mxnet/pull/12157#discussion_r211713108

Reply:
Because in this PR,  we allow to register same backend multiple times, and also possible to register same property multiple times. `__REGISTER__` won't cause terminating anymore.

Reply:
Removing the check on line 181: https://github.com/apache/incubator-mxnet/pull/14277/files#diff-4db2c685581a8f376abff8d6ac908e1cL181 will not prevent a subgraph property from registering more than once. I think the intention was to create only one shared pointer to one subgraph property object and use it everywhere. We should stick to doing this. One way is to modify the subgraph registry to create unordered_map<string, map<string, unoredered_map<string, createfunc>> . Then you register each subgraph property with top level name and second level name to search within the inner map.

Reply:
Basically, the main change here is, replace `std::unordered_map<std::string, SubgraphPropertyCreateFn>` with `std::unordered_map<std::string, std::vector<SubgraphPropertyCreateFn>>`  as the type of prop_fn_map_. 
Using `std::vector` instead of `unoredered_map`  as the inner type because, we need to keep `SubgraphPropertyCreateFn` in a user defined order, while unoredered_map can't guarantee it.

Reply:
@ZhennanQin Here is an example: https://github.com/anirudh2290/mxnet/tree/fix_bind . I created an additional.h which has a new subgraph property (test) and selector. It is included from multiple source files and thus there are multiple subgraph properties in prop_fn["test"]. Partition Graph is also called twice instead of once. Every unique subgraph property should register only once.

Reply:
I understand your point.  As quantized FC is merged into master. I rebased this PR to show what we needed here:
https://github.com/apache/incubator-mxnet/blob/f3b5e980581c94c9d36b5a9f02cbb923be9ba7bd/src/operator/subgraph/mkldnn/mkldnn_subgraph_property.cc#L30-L34

In our design, we will register same backend name many times with different subgraph properties.  Property will run inthe order they registered. 

It's possible to register unique subgraph property many times for different backends or even for unique backend. Below registration is possible in future:

```
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN, SgMKLDNNConvProperty);
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN, SgMKLDNNFCProperty);
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN_POST_QUANTIZE, SgMKLDNNConvProperty);
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN_POST_QUANTIZE, SgMKLDNNFCProperty);
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN_POST_QUANTIZE, SgMKLDNNConvPostQuantizeProperty);
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN_POST_QUANTIZE, SgMKLDNNFCPostQuantizeProperty);
```
As you seen, `SgMKLDNNConvProperty` is registered twice for different backend names.

Also, it's possible to have this in future:
```
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN, SgMKLDNNConvProperty);
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN, SgMKLDNNFCProperty);
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN, SgMKLDNNConversionProperty);
MXNET_REGISTER_SUBGRAPH_PROPERTY(MKLDNN, SgMKLDNNConvProperty);
```
We need to register `SgMKLDNNConvProperty` twice within `MKLDNN`, and we need to run each property as the order registered. The reason might be, `SgMKLDNNConversionProperty` will do further optimization on fused conv, and this conversion might expose more opportunities for `SgMKLDNNConvProperty`, so we need to run `SgMKLDNNConvProperty` again.

As backend implements more and more subgraph pass, it's highly possible to have above requirement. We need to design for it in advance.

Reply:
In my code, I use MXNET_REGISTER_SUBGRAPH_PROPERTY only once here: https://github.com/anirudh2290/mxnet/blob/fix_bind/src/operator/subgraph/additional.h#L66 . But include in two different source files( https://github.com/anirudh2290/mxnet/blob/fix_bind/src/operator/subgraph/default_subgraph_property.cc#L23 , https://github.com/anirudh2290/mxnet/blob/fix_bind/src/operator/subgraph/mkldnn/mkldnn_subgraph_property.cc#L24) which will be part of different translation units. As you can see, it will create two properties and run PartitionGraph twice. This is not intended and different from your example where you have explicitly called MXNET_REGISTER_SUBGRAPH_PROPERTY twice with the same name and property.

Reply:
@anirudh2290 Putting MXNET_REGISTER_SUBGRAPH_PROPERTY in header file is not a good usecase. If developer include that header twice, then I think he just wants to register that properties twice for same backend. I admited this may cause misleading if developer doesn't know the details behind MXNET_REGISTER_SUBGRAPH_PROPERTY . But I have no idea how to avoid this. Do you have any good suggestion? Please consider the requirement I listed in previous comment.

Reply:
We can have a different macro for registering property versus registering backend. For example:
```
MXNET_REGISTER_SUBGRAPH_PROPERTY(DefaultSubgraphProperty)
MXNET_REGISTER_BACKEND(default, DefaultSubgraphProperty, DefaultSubgraphProperty)
```

I don't mind sticking to the limitation of including MXNET_REGISTER_SUBGRAPH_PROPERTY to only from .cc files but we need to document this well otherwise this can lead to difficult to trace bugs.

Reply:
@anirudh2290 I've consider the way you suggested, including other formats like,
```
MXNET_REGISTER_BACKEND(default, {DefaultSubgraphProperty, DefaultSubgraphProperty})
```
or
```
MXNET_REGISTER_BACKEND(default, DefaultSubgraphProperty, 0)
MXNET_REGISTER_BACKEND(default, DefaultSubgraphProperty, 1)
```
But I give them up because of different reasons. The first problem is, `DefaultSubgraphProperty` is only the name of class, but not a object or function. So it can't be inserted in a vector. The correct line is,
```
MXNET_REGISTER_BACKEND(default, {DefaultSubgraphProperty::Create(), DefaultSubgraphProperty::Create()})
```
But it's not elegant.

For your suggestion,
```
MXNET_REGISTER_BACKEND(default, DefaultSubgraphProperty, DefaultSubgraphProperty)
```
It should work if we define `MXNET_REGISTER_BACKEND` with 3 arguments. But it's not scalable because if we want to register more property for this backend, we need to write more `MXNET_REGISTER_BACKEND` implementation with more number of arguments.

For
```
MXNET_REGISTER_BACKEND(default, DefaultSubgraphProperty, 0)
MXNET_REGISTER_BACKEND(default, DefaultSubgraphProperty, 1)
```
This can help to avoid register property twice unexceptly from header as it defines the execution order. But this will make code a bit ugly.

Personally, I prefer the current implementation although we need to use it carefully, like not putting `MXNET_REGISTER_SUBGRAPH_PROPERTY` in header. But I don't mind to change to other idea I listed above. Then you can make the final decision now, and I will follow it.

Reply:
I think I am okay with the REGISTER only in cc limitation as long as it is well documented here: http://mxnet.incubator.apache.org/versions/master/tutorials/c++/subgraphAPI.html?highlight=subgraph#subgraph-api . 

Also, we need to make sure that we call Create only once for each property even for this use case:
```
MXNET_REGISTER_SUBGRAPH_PROPERTY(default, DefaultSubgraphProperty)
MXNET_REGISTER_SUBGRAPH_PROPERTY(default, DefaultSubgraphProperty)
```
This can be done by adding a map of subgraph property name to the corresponding subgraph property pointer. For this passing the property name along with the create fn is required when calling __REGISTER__. The create gets called and returned subgraph property ptr gets added in __REGISTER__ itself. We can then reuse the already added subgraph property ptr in CreateSubgraphProperty.

Reply:
Code is refactored as your comments. Please review again. Thanks.